cpp/src/interop/to_arrow_host.cu

```diff
@@ -0,0 +1,426 @@
+
+  NANOARROW_RETURN_NOT_OK(populate_validity_bitmap(ArrowArrayValidityBitmap(tmp.get())));
+  auto bitmask = bools_to_mask(column, stream, mr);
+  NANOARROW_RETURN_NOT_OK(populate_data_buffer(
+    device_span<uint8_t const>(reinterpret_cast<const uint8_t*>(bitmask.first->data()),
+                               bitmask.first->size()),
+    ArrowArrayBuffer(tmp.get(), fixed_width_data_buffer_idx)));
+
+  ArrowArrayMove(tmp.get(), out);
+  return NANOARROW_OK;
+}
+
+template <>
+int dispatch_to_arrow_host::operator()<cudf::string_view>(ArrowArray* out) const
+{
+  ArrowType nanoarrow_type = NANOARROW_TYPE_STRING;
+  if (column.num_children() > 0 &&
+      column.child(cudf::strings_column_view::offsets_column_index).type().id() ==
+        cudf::type_id::INT64) {
+    nanoarrow_type = NANOARROW_TYPE_LARGE_STRING;
+  }
+
+  nanoarrow::UniqueArray tmp;
+  NANOARROW_RETURN_NOT_OK(initialize_array(tmp.get(), nanoarrow_type, column));
+
+  if (column.size() == 0) {
+    // initialize the offset buffer with a single zero by convention
+    if (nanoarrow_type == NANOARROW_TYPE_LARGE_STRING) {
+      NANOARROW_RETURN_NOT_OK(
+        ArrowBufferAppendInt64(ArrowArrayBuffer(tmp.get(), fixed_width_data_buffer_idx), 0));
+    } else {
+      NANOARROW_RETURN_NOT_OK(
+        ArrowBufferAppendInt32(ArrowArrayBuffer(tmp.get(), fixed_width_data_buffer_idx), 0));
+    }
+
+    ArrowArrayMove(tmp.get(), out);
+    return NANOARROW_OK;
+  }
+
+  NANOARROW_RETURN_NOT_OK(populate_validity_bitmap(ArrowArrayValidityBitmap(tmp.get())));
+
+  auto const scv     = cudf::strings_column_view(column);
+  auto const offsets = scv.offsets();
+  if (offsets.type().id() == cudf::type_id::INT64) {
+    NANOARROW_RETURN_NOT_OK(populate_data_buffer(
+      device_span<int64_t const>(offsets.data<int64_t>() + scv.offset(), scv.size() + 1),
+      ArrowArrayBuffer(tmp.get(), fixed_width_data_buffer_idx)));
+  } else {
+    NANOARROW_RETURN_NOT_OK(populate_data_buffer(
+      device_span<int32_t const>(offsets.data<int32_t>() + scv.offset(), scv.size() + 1),
+      ArrowArrayBuffer(tmp.get(), fixed_width_data_buffer_idx)));
+  }
+
+  NANOARROW_RETURN_NOT_OK(
+    populate_data_buffer(device_span<char const>(scv.chars_begin(stream), scv.chars_size(stream)),
+                         ArrowArrayBuffer(tmp.get(), 2)));
+
+  ArrowArrayMove(tmp.get(), out);
+  return NANOARROW_OK;
+}
+
+template <>
+int dispatch_to_arrow_host::operator()<cudf::list_view>(ArrowArray* out) const
+{
+  nanoarrow::UniqueArray tmp;
+  NANOARROW_RETURN_NOT_OK(initialize_array(tmp.get(), NANOARROW_TYPE_LIST, column));
+  NANOARROW_RETURN_NOT_OK(ArrowArrayAllocateChildren(tmp.get(), 1));
+
+  NANOARROW_RETURN_NOT_OK(populate_validity_bitmap(ArrowArrayValidityBitmap(tmp.get())));
+  auto const lcv = cudf::lists_column_view(column);
+  if (column.size() == 0) {
+    NANOARROW_RETURN_NOT_OK(
+      ArrowBufferAppendInt32(ArrowArrayBuffer(tmp.get(), fixed_width_data_buffer_idx), 0));
```

Comment:
Could we add the same comment you included for strings above about the empty offsets case? If there's any resource we could link to that would be ideal, but IIRC we discussed this last time and it's largely undocumented convention at this stage (correct me if I'm wrong, though).